### PR TITLE
[refactor] moving grad acc logic to engine

### PR DESCRIPTION
**Pull request by feifeibear**
Avoid circle import.

### CODE REVIEW

**Comment by feifeibear** [comment]
The failed CI case is due to address in use. I tested it locally.
image

**Comment by FrankLeeeee** [comment]
> 

Ok, this occurs as the `rerun_if_address_in_use` decorator is not added.

File: colossalai/engine/schedule/_pipeline_schedule.py [hunk 1]
```diff
@@ -386,8 +387,7 @@ def __init__(self,
         self.num_model_chunks = num_model_chunks
 
     def pre_processing(self, engine):
-        # FIXME(jiaruifang) we shall not use ShardedModelV2 in pipeline mode, due to circular dependency.
-        if hasattr(engine.model, 'colo_attr'):
+        if isinstance(engine.model[0], ShardedModelV2):
```
Comment (ver217) [inline colossalai/engine/schedule/_pipeline_schedule.py:390]: Maybe we can use `isinstance(engine.model[0], (ShardedModelV2, NaiveAMPModel))`

Comment (feifeibear) [reply colossalai/engine/schedule/_pipeline_schedule.py]: This is a bug. It should be if isinstance(engine.model, ShardedModelV2):
I fixed it

**Review by ver217 — APPROVED** [review]

**Pull request merged**